core/src/main/scala/spinal/core/formal/FormalBootstraps.scala

```diff
@@ -272,3 +290,28 @@ case class SpinalFormalConfig(
     }
   }
 }
+
+class PhaseFormalVhdlFixups extends PhaseNetlist {
```

Comment:
I understand the usage of this part, how ever IMO this class is best been put into core\src\main\scala\spinal\core\internals\PhaseVhdl.scala not in FormalBootstraps.

Reply:
What a nice idea!
With the new commits, I moved the code to the `ComponentEmitterVhdl.scala` to the various other _preprocessing_ steps in `emitArchitecture()`. It converts tha assignments to tags/attributes which the rest of the component emitter correctly emits, and then goes on to ignore/skip the assignment in `emitAssignment()`.

Reply:
So may be the steps remains should be add test cases in the existing verification tester?
you can search them by using FormalConfig keyword.

Reply:
I already have many test cases that test various conditions and functionalities for correctness i.e. equivalence between SystemVerilog and VHDL generation... uploading them (as I'm heavily editing them) is not good right now, especially since the CI runner will fail anyway because https://github.com/SpinalHDL/docker/pull/11 is not merged yet.

The biggest issue I currently have, is that the PSL asserts cannot live within a conditional/process. I.e. when I write:
```scala
when(False) {
  assert(False)
}
```

The assert should obviously never trigger. In Verilog it generates:
```systemverilog
  assign when_FormalApi_l379 = 1'b0;
  always @(posedge clk or posedge reset) begin
    if(reset) begin
    end else begin
      if(when_FormalApi_l379) begin
        assert(1'b0); // FormalApi.scala:L381
      end
    end
  end
```

but for VHDL it currently only generates (manually trimmed):

```vhd
architecture arch of AssertWithinCondTest is
begin
  FormalApi_L381: assert always (pkg_toStdLogic(false) = '1') abort (reset = '1') @rising_edge(clk); -- FormalApi.scala:L381
end arch;
```

So right now I'm trying to figure out how I can walk backwards the data model. I'm at the `AssertStatement` and need to figure out in which conditional scopes this statement is, up to the root level. With this info I could then generate the necessary conditionals and AND it to the conditional of the assert itself.

Maybe @Dolu1990 can help how to walk the conditional scopes backwards/up?

Reply:
ChatGPT tell me that it is
``` VHDL
property when_FormalApi_l379_implication is always (
    when_FormalApi_l379 -> next (FALSE)
);
```
If so, is it possible to convert it to be
``` VHDL
FormalApi_L381: assert always (when_FormalApi_l379 -> next (FALSE) abort (reset = '1') @rising_edge(clk);
```

Reply:
> What a nice idea! With the new commits, I moved the code to the `ComponentEmitterVhdl.scala` to the various other _preprocessing_ steps in `emitArchitecture()`. It converts tha assignments to tags/attributes which the rest of the component emitter correctly emits, and then goes on to ignore/skip the assignment in `emitAssignment()`.

After having a conversation with @Dolu1990 , he thinks editing emit for formal verification code would not be a good practice then using a standalone phase.

Reply:
> After having a conversation with @Dolu1990 , he thinks editing emit for formal verification code would not be a good practice then using a standalone phase.

I don't know if I understand this sentence correctly: Is he saying that it would be better to not handle formal stuff in `ComponentEmitterVhdl.scala` at all? So better would be to handle in in a new phase?

Changes to `ComponentEmitterVhdl.scala` are required anyway as it does not know how to handle `Operator.Formal.<x>` yet. The Verilog/SystemVerilog emitter does know.

IMHO we could later always move out the formal specific stuff to a separate phase i.e. if it becomes too big. Currently it is only that small part that dies the transformation of random assignments to attributes.

Reply:
> ```vhdl
> FormalApi_L381: assert always (when_FormalApi_l379 -> next (FALSE) abort (reset = '1') @rising_edge(clk);
> ```
Yes, with properties and temporal sequences is how this can be done. A naive implementation could also just do:
```vhdl
FormalApi_L381: assert always ((not when_FormalApi_l379) or FALSE) abort (reset = '1') @rising_edge(clk);
```
But handling it with properties i.e. breaking it up in multiple lines is better for readability.

Reply:
> > After having a conversation with @Dolu1990 , he thinks editing emit for formal verification code would not be a good practice then using a standalone phase.
> 
> I don't know if I understand this sentence correctly: Is he saying that it would be better to not handle formal stuff in `ComponentEmitterVhdl.scala` at all? So better would be to handle in in a new phase?
> 
> Changes to `ComponentEmitterVhdl.scala` are required anyway as it does not know how to handle `Operator.Formal.<x>` yet. The Verilog/SystemVerilog emitter does know.
> 
> IMHO we could later always move out the formal specific stuff to a separate phase i.e. if it becomes too big. Currently it is only that small part that does the transformation of random assignments to attributes.

You can have a look at #1435 as reference.

Reply:
> 
> But handling it with properties i.e. breaking it up in multiple lines is better for readability.

agree，but we can implement it step by step.

maybe after define a FormalProperty traits first.

Reply:
The newest commits did move the formal phase fixups to the new `FormalPhase` from #1435. I also managed to unroll the scope a statement is in to build the conditional for it. This is then used as a preface in the PSL property (with `|->`) such that conditional assume and asserts are possible now.

Reply:
Do you think we should use the same symbol with SVA and PSL? 
The property has antecedent and consequent which act like `when() othewise` block.
Should we create a new syntax sugar for that?

Reply:
Which symbol do you refer to? Are you suggesting to also use the `|->` operator for SVA? I don't think that would be good as the open version of Yosys does not support that.

My current goal is simply to reach feature equivalence of verilog and vhdl backends. 

To extend that with the advanced features (temporal statements, properties etc.) I plan on doing in a future but separate PR. Step by step. :)

Reply:
> To extend that with the advanced features (temporal statements, properties etc.) I plan on doing in a future but separate PR. Step by step. :)

Nice plan. the test failed now is because of the CI environment. I think we need to upgrade it first.
How about upgrade the docker image with ghdl and related plugin only? First.